pkg/api/util/names.go

```diff
@@ -44,15 +44,65 @@ func ComputeName(prefix string, obj interface{}) (string, error) {
 	// and pods down the road for ACME resources.
 	prefix = DNSSafeShortenTo52Characters(prefix)
 
+	// the prefix is <= 52 characters, the decimal representation of
+	// the hash is <= 10 characters, and the hyphen is 1 character.
+	// 52 + 10 + 1 = 63, so we're good.
 	return fmt.Sprintf("%s-%d", prefix, hashF.Sum32()), nil
 }
 
-// DNSSafeShortenTo52Characters shortens the input string to 52 chars and ensures the last char is an alpha-numeric character.
-func DNSSafeShortenTo52Characters(in string) string {
-	if len(in) >= 52 {
-		validCharIndexes := regexp.MustCompile(`[a-zA-Z\d]`).FindAllStringIndex(fmt.Sprintf("%.52s", in), -1)
-		in = in[:validCharIndexes[len(validCharIndexes)-1][1]]
+// ComputeSecureUniqueDeterministicNameFromData computes a deterministic name from the given data.
+// The algorithm in use is SHA256 and is cryptographically secure.
+// The output is a string that is safe to use as a DNS label.
+// The output is guaranteed to be unique for the given input.
+// The output will be at least 64 characters long.
+func ComputeSecureUniqueDeterministicNameFromData(fullName string, maxNameLength int) (string, error) {
+	const hashLength = 64
+	if maxNameLength < hashLength {
+		return "", fmt.Errorf("maxNameLength must be at least %d", hashLength)
+	}
+
+	if len(fullName) <= maxNameLength {
+		return fullName, nil
+	}
+
+	hash := sha256.New()
+
+	_, err := hash.Write([]byte(fullName))
+	if err != nil {
+		return "", err
+	}
+
+	prefix := DNSSafeShortenToNCharacters(fullName, maxNameLength-hashLength-1)
```

Comment:
`fullName` is already a DNS subdomain, and we want another shorter DNS subdomain. Why is `DNSSafeShortenToNCharacters` required here?

Ah, that's because trimming the end of a DNS subdomain may yield a non-DNS subdomain. Example:

```diff
-super-mega-long-certificate-name-that-uses-a-ton-of-characters
+super-mega-long-certificate-name-that-uses-a-ton-of-
 <-------------------------------------------------->
                       52 chars                     ^
                                                not a DNS label
                                                due to the ending
                                                "-"
```

It would be good to explain that here, e.g.,

```suggestion
	// Although fullName is already a DNS subdomain, we can't just cut it
	// at 52 characters and expect another DNS subdomain. That's because
	// we might cut it right after a "-" or a ".", which would give an
	// invalid DNS subdomain.
	prefix := DNSSafeShortenToNCharacters(fullName, maxNameLength-hashLength-1)
```

Reply:
Ah, that's because trimming the end of a DNS subdomain may yield a non-DNS subdomain. Example:

```diff
-super-mega-long-certificate-name-that-uses-a-ton-of-characters
+super-mega-long-certificate-name-that-uses-a-ton-of-
 <-------------------------------------------------->
                       52 chars                     ^
                                                not a DNS label
                                                due to the ending
                                                "-"
```

It would be good to explain that here, e.g.,

```suggestion
	// Although fullName is already a DNS subdomain, we can't just cut it
	// at 52 characters and expect another DNS subdomain. That's because
	// we might cut it right after a "-" or a ".", which would give an
	// invalid DNS subdomain.
	prefix := DNSSafeShortenToNCharacters(fullName, maxNameLength-hashLength-1)
```

Reply:
We still add a "-<hash>", so it might actually yield a valid dns name, not a beautiful one (eg. "test--<hash>")
But it could indeed be an invalid name too (eg. "test.-<hash>"

Reply:
I updated the PR, @maelvls PTAL